### DISCUSSION TITLE
How do I pass options to a custom field type?

### DISCUSSION THREAD

**Discussion by omarish** [pydantic/pydantic#3253] [Question]
Hi, I had a question about how to make custom fields that I can configure.

Example:

```python
class EthereumAddress(str):
    # There's some custom validation logic that works here
    ...
```

I'd like to be able to pass a param in to this field that would affect how validation runs:

```python
class Model(pydantic.BaseModel):
    address: EthereumAddress(allow_ens=True)
```

I was thinking of implementing `__new__` the `EthereumAddress` class, but was wondering if there was a more straightforward way to do this.

Thanks.

**Comment by jayqi** [comment]
I was able to get something to work using a custom class that implements `__get_pydantic_core_schema__` that can be attached to my source type with `typing.Annotated`. 

Here is an example of how it gets used. In my case, `SortedSet` is like your `EthereumAddress`. `SortedSet` has an `__init__` keyword argument `key`. I'm trying to pass in `key=lambda x: -x`, like your `allow_ens=True`. 

```python
from typing import Annotated

from pydantic import BaseModel
from sortedcontainers_pydantic import Key, SortedSet

class MyModel(BaseModel):
    sorted_set: Annotated[SortedSet[int], Key(lambda x: -x)]

MyModel(sorted_set=[3.0, 1.0, 2.0])
#> MyModel(sorted_set=SortedSet([3, 2, 1], key=<function MyModel.<lambda> at 0x1089b3060>))
```

<sup>Created at 2025-04-18 03:00 EDT by [reprexlite](https://github.com/jayqi/reprexlite) v1.0.0</sup>


You can see my implementation source code [here](https://github.com/drivendataorg/sortedcontainers-pydantic/blob/5d76627dd247e76f94b1a487c9cd2ebf536e4642/sortedcontainers_pydantic/__init__.py#L321-L343). 

The main ideas:

- `Key` is a class that implements `__get_pydantic_core_schema__`, which means that you can attach it with `Annotated` and it will get picked up by the Pydantic validation machinery. This is inspired by the `MyAfterValidator` example in the [Pydantic docs](https://docs.pydantic.dev/latest/concepts/types/#as-an-annotation). 
- You pass in your parameters as `__init__` arguments on `Key`
- You can check the `source_type` to make sure it's a type you expect. The first argument of `Annotated` gets passed in here. 
- You can use `functools.partial` to make a constructor callable to give to the validator that includes your parameters.
